javascript/angular-todo-davinci/src/app/services/davinci/davinci.service.ts

```diff
@@ -0,0 +1,167 @@
+/*
+ * ping-sample-web-angular-davinci
+ *
+ * davinci.service.ts
+ *
+ * Copyright (c) 2025 Ping Identity Corporation. All rights reserved.
+ * This software may be modified and distributed under the terms
+ * of the MIT license. See the LICENSE file for details.
+ */
+
+import { computed, Injectable, Signal, signal, WritableSignal } from '@angular/core';
+import createClient from './davinci.utils';
+import { DaVinciClient, DaVinciNode } from './davinci.types';
+import {
+  Collectors,
+  FlowCollector,
+  PasswordCollector,
+  TextCollector,
+  Updater,
+  ValidatedTextCollector,
+} from '@forgerock/davinci-client/types';
+
+/**
+ * Used to manage DaVinci flows
+ */
+@Injectable()
+export class DavinciService {
+  /***********************************************************************
+   * SDK INTEGRATION POINT
+   * Summary: Create state needed to render the form
+   * ----------------------------------------------------------------------
+   * Details: Whenever the node changes we will get the updated client
+   * info and extract data needed to render the form. We take advantage of
+   * computed values to prevent unnecessary re-renders of the form when the
+   * node and collectors change.
+   ************************************************************************* */
+  private client: WritableSignal<DaVinciClient | null> = signal(null);
+  readonly node: WritableSignal<DaVinciNode | null> = signal(null);
+  collectors: Signal<Collectors[]> = computed(() => {
+    const currentNode = this.node();
+    if (currentNode?.status === 'continue') {
+      return this.client().getCollectors() ?? [];
+    } else return [];
+  });
+  formName: Signal<string> = computed(() => {
+    const currentNode = this.node();
+    if (currentNode?.status === 'continue') {
+      return currentNode.client.name ?? '';
+    } else {
+      return '';
+    }
+  });
+  formAction: Signal<string> = computed(() => {
+    const currentNode = this.node();
+    if (currentNode?.status === 'continue') {
+      return currentNode.client.action ?? '';
+    } else {
+      return '';
+    }
+  });
+  errorMessage: Signal<string> = computed(() => {
```

Comment:
I think a lot of these helper functions that you have written here should be no longer needed if you use our "selector" functions, rather than using regular dot notation to retrieve object properties. Our API has `getClient`, `getError`, `getCollectors` ... all of these handle `null` values and return the correct types for you. If you really leverage these selectors, you shouldn't need all these checks.

One additional thing, as Ryan mentioned elsewhere, is that if we do type-narrowing correctly, and components are only executed one the type is guaranteed, TS won't error if you use properties that are only present in certain status types.

Reply:
Part of the reason why I didn't use `getClient` is because it didn't have nice return types I could use so I would have to infer the type like `ReturnType<DaVinciClient['getClient']>`. But I can make that update.

While I can get rid of the computed values for `collectors` and `errorMessage`, I'm not seeing how I can get rid of `formName` and `formAction` since I need to check for a continue status still in order to get those properties.